idaes/generic_models/properties/core/reactions/tests/test_equilibrium_forms.py

```diff
@@ -324,3 +326,203 @@ def test_log_power_law_equil_with_order():
+    assert m.rparams.reaction_r1.reaction_order["sol", "c1"].value == 5
+    assert m.rparams.reaction_r1.reaction_order["sol", "c2"].value == 6
+
+    # Check reaction form
+    rform = solubility_product.return_expression(
+        m.rxn[1], m.rparams.reaction_r1, "r1", 300)
+
+    s = (m.thermo[1].mole_frac_phase_comp["sol", "c1"] +
+         m.thermo[1].mole_frac_phase_comp["sol", "c2"])
+    Q = (m.rxn[1].k_eq["r1"] -
+         (m.thermo[1].mole_frac_phase_comp["p1", "c1"] **
+          m.rparams.reaction_r1.reaction_order["p1", "c1"] *
+          m.thermo[1].mole_frac_phase_comp["p1", "c2"] **
+          m.rparams.reaction_r1.reaction_order["p1", "c2"] *
+          m.thermo[1].mole_frac_phase_comp["p2", "c1"] **
+          m.rparams.reaction_r1.reaction_order["p2", "c1"] *
+          m.thermo[1].mole_frac_phase_comp["p2", "c2"] **
+          m.rparams.reaction_r1.reaction_order["p2", "c2"] *
+          m.thermo[1].mole_frac_phase_comp["sol", "c1"] **
+          m.rparams.reaction_r1.reaction_order["sol", "c1"] *
+          m.thermo[1].mole_frac_phase_comp["sol", "c2"] **
+          m.rparams.reaction_r1.reaction_order["sol", "c2"]))
+
+    assert str(rform) == str(
+        s - smooth_max(0, s-Q, m.rparams.reaction_r1.eps) == 0)
+
+
+@pytest.mark.unit
+def test_solubility_no_solids():
+    m = ConcreteModel()
+
+    # # Add a test thermo package for validation
+    m.pparams = PhysicalParameterTestBlock()
+
+    m.thermo = m.pparams.build_state_block([1])
+
+    # Create a dummy reaction parameter block
+    m.rparams = GenericReactionParameterBlock(default={
+        "property_package": m.pparams,
+        "base_units": {"time": pyunits.s,
+                       "mass": pyunits.kg,
+                       "amount": pyunits.mol,
+                       "length": pyunits.m,
+                       "temperature": pyunits.K},
+        "equilibrium_reactions": {
+            "r1": {"stoichiometry": {("p1", "c1"): -1,
+                                     ("p1", "c2"): 2},
+                   "equilibrium_form": solubility_product,
+                   "concentration_form": ConcentrationForm.moleFraction}}})
+
+    # Create a dummy state block
+    m.rxn = Block([1])
+    add_object_reference(
+        m.rxn[1], "phase_component_set", m.pparams._phase_component_set)
+    add_object_reference(m.rxn[1], "params", m.rparams)
+    add_object_reference(m.rxn[1], "state_ref", m.thermo[1])
+
+    m.rxn[1].k_eq = Var(["r1"], initialize=1)
+
+    solubility_product.build_parameters(
+        m.rparams.reaction_r1,
+        m.rparams.config.equilibrium_reactions["r1"])
+
+    # Check reaction form - should raise exception
+    with pytest.raises(ConfigurationError,
+                       match="did not find a solid phase component for "
+                       "precipitation reaction r1. This is likely due to the "
+                       "reaction configuration."):
+        solubility_product.return_expression(
```

Comment:
Does this mean that for a subsaturated solution, a config error will also be generated, or is this only because solids were omitted from the reaction parameter block config? 

I'm guessing this was written from the perspective of triggering precipitation reactions for pretreatment processes, i.e., with the main focus placed on the presence of solids. I am just thinking about how we could implement this to verify that a solution is still undersaturated with respect to some potential foulant (e.g., in the case where we want to avoid mineral scaling on desalination membranes).

Reply:
This check is only looking for the definition of a solid phase component, not whether there is any actually present in the system.

The model formulation is designed to support sub-saturated systems, and that is part of why it is as complex as it is.

Reply:
From `If S = 0, Q >= 0`, the undersaturated case should be handled, so I guess the user just has to be sure to define the potential solids in the reaction config. Is that correct?

Reply:
Yes, that is correct. The error is to just catch the case where the potential solid was not defined, in which case `S=None` which won't work out well in an expression.